### ISSUE TITLE
NaNs in loss function

### ISSUE THREAD

**Issue by jfsantos** [johnglover/sound-rnn#2]
I am currently trying to find the reason for the NaNs in the loss function. I found the guilty already, which are zero values in the `norm` variable inside the function `pdf`, which then causes divisions by zero. A no-brainer workaround would be to simply add a small value to both the variances and the norms, but maybe you have a better idea?

I'm getting these NaNs really soon (around the 500th iteration) in training when using the GPU (which might be due to the float16 representation) and mini-batches of 64 samples. I don't get them when using the fix I proposed, so it might be worth adding it to the repository. I can submit a PR if you think this fix is acceptable.

**Comment by johnglover** [comment]
Hmm. I haven't tried running this on GPUs yet, hopefully I'll get a chance over the weekend. So your variances are becoming zero here?

**Comment by jfsantos** [comment]
Yes, that's what is happening. There's nothing to avoid having zeros in the variances, so when it happens it will simply make the loss NaN, from which training will never be able to recover. I tried adding 1e-12 to the variance vector and it seemed to work well with a few models I've trained (both on the sin data and some other audio files).

**Comment by johnglover** [comment]
Ok thanks, sounds reasonable, so you're adding 1e-12 after the values have been passed through the exp function I assume? Happy to merge that if you don't mind making a PR.

**Comment by johnglover** [comment]
I made this change in the above commit, thanks.

**Comment by quantumdude836** [comment]
Similar issue is happening with test.lua - some of the elements of `num_grad` are zero, so `err` winds up with `nan`, causing the test to fail.
